Missing unifications with never type

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn foo() {}

enum Foo {
    A,
    B,
}

fn main() {
    let a = Foo::A;

    loop {
        match a {
            Foo::A => break,
            Foo::B => foo(),
        }
        match a {
            Foo::B => foo(),
            Foo::A => break,
        }
    }
}

```

I expected to see this happen: no error. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1b488957709fc4a91b99fee0fb4ba3dd

Instead, this happened: two type mismatch errors.

```rust
expect-never.rs:12:9: error: mismatched types, expected ‘!’ but got ‘()’ [E0308]
   12 |         match a {
      |         ^~~~~
   13 |             Foo::A => break,
      |                       ~~~~~
   14 |             Foo::B => foo(),
      |                       ~~~
expect-never.rs:12:9: error: mismatched types, expected ‘()’ but got ‘<tyty::error>’ [E0308]
   12 |         match a {
      |         ^~~~~
expect-never.rs:16:9: error: mismatched types, expected ‘()’ but got ‘!’ [E0308]
   16 |         match a {
      |         ^~~~~
   17 |             Foo::B => foo(),
      |                       ~~~
   18 |             Foo::A => break,
      |                       ~~~~~
expect-never.rs:11:10: error: expected ‘()’ got <tyty::error>
   11 |     loop {
      |          ^
```

This is due to missing unification in `rust-unify.cc` I believe? For example, `expect_never` will error out if the given `rtype` is not `TyTy::NEVER` too, but I believe it should unify/resolve to the `ltype` in all cases.

Similarly, in the cases where we have almost any type in `ltype` and a `TyTy::NEVER` in `rtype`, we should resolve to `ltype` in all cases.

--- a/gcc/rust/typecheck/rust-unify.cc
+++ b/gcc/rust/typecheck/rust-unify.cc
@@ -17,6 +17,7 @@
 // <http://www.gnu.org/licenses/>.
 
 #include "rust-unify.h"
+#include "rust-tyty.h"
 
 namespace Rust {
 namespace Resolver {
@@ -236,6 +237,15 @@ UnifyRules::go ()
 	  ltype = i;
 	}
     }
+
+  // The never type should always get coerced to the type it's being matched
+  // against, so in that case, ltype. This avoids doing the same check in all
+  // the `expect_*` functions.
+  // However, this does not work if we have an annoying ltype - like INFER.
+  // TODO: Is ltype == Infer the only special case here? What about projections?
+  // references?
+  if (rtype->get_kind () == TyTy::NEVER && ltype->get_kind () != TyTy::INFER)
+    return ltype->clone ();
 
   switch (ltype->get_kind ())
     {
@@ -1536,32 +1546,8 @@ UnifyRules::expect_never (TyTy::NeverType *ltype, TyTy::BaseType *rtype)
       }
       break;
 
-    case TyTy::NEVER:
+    default:
       return rtype->clone ();
-
-    case TyTy::PLACEHOLDER:
-    case TyTy::PROJECTION:
-    case TyTy::DYNAMIC:
-    case TyTy::CLOSURE:
-    case TyTy::SLICE:
-    case TyTy::PARAM:
-    case TyTy::POINTER:
-    case TyTy::STR:
-    case TyTy::ADT:
-    case TyTy::REF:
-    case TyTy::ARRAY:
-    case TyTy::FNDEF:
-    case TyTy::FNPTR:
-    case TyTy::TUPLE:
-    case TyTy::BOOL:
-    case TyTy::CHAR:
-    case TyTy::INT:
-    case TyTy::UINT:
-    case TyTy::FLOAT:
-    case TyTy::USIZE:
-    case TyTy::ISIZE:
-    case TyTy::ERROR:
-      return new TyTy::ErrorType (0);
     }
   return new TyTy::ErrorType (0);
 }

--- a/gcc/testsuite/rust/compile/match-never-ltype.rs
+++ b/gcc/testsuite/rust/compile/match-never-ltype.rs
@@ -1,0 +1,17 @@
+fn foo() {}
+
+enum Foo {
+    A,
+    B,
+}
+
+fn main() {
+    let a = Foo::A;
+
+    loop {
+        match a {
+            Foo::A => break,
+            Foo::B => foo(),
+        }
+    }
+}

--- a/gcc/testsuite/rust/compile/match-never-rtype.rs
+++ b/gcc/testsuite/rust/compile/match-never-rtype.rs
@@ -1,0 +1,17 @@
+fn foo() {}
+
+enum Foo {
+    A,
+    B,
+}
+
+fn main() {
+    let a = Foo::A;
+
+    loop {
+        match a {
+            Foo::B => foo(),
+            Foo::A => break,
+        }
+    }
+}

